File: src/infrastructure/storage/board_persistence.py
from pathlib import Path
from src.core.types import ItemId
from src.core.constants import BOARD_FILENAME, COLUMN_METADATA_FILENAME
from src.utils.file_utils import (
    ensure_directory_exists,
    safe_delete_file,
    safe_rename_file,
    get_unique_filename,
)
from src.utils.string_utils import get_title_filename
from src.utils.date_utils import now
from src.infrastructure.storage.file_operations import (
    find_item_file_by_id,
    get_board_directory_path,
    get_column_directory_path,
    cleanup_column_files,
)
from src.infrastructure.storage.markdown_parser import (
    save_item_with_metadata,
    save_column_metadata,
)


class BoardPersistence:
    def __init__(self, data_dir: Path):
        self.data_dir = Path(data_dir)
        ensure_directory_exists(self.data_dir)

        self.boards_dir = self.data_dir / "boards"
        ensure_directory_exists(self.boards_dir)

    def save_item_to_column(
        self, board_name: str, column_name: str, item_data: dict
    ) -> None:
        # Add debug logging
        import logging
        logger = logging.getLogger("mkanban-daemon")
        logger.debug(f"save_item_to_column: board='{board_name}', column='{column_name}', item='{item_data.get('title', 'unknown')}'")

        board_dir = get_board_directory_path(self.boards_dir, board_name)
        column_dir = get_column_directory_path(board_dir, column_name)
        logger.debug(f"Column directory: {column_dir}")
        ensure_directory_exists(column_dir)

        item_id = item_data["id"]
        title = item_data["title"]
        content = item_data.get("description", "")

        new_filename = get_title_filename(title)
        old_item_file = find_item_file_by_id(column_dir, item_id)
        new_item_file = column_dir / f"{new_filename}.md"

        if old_item_file and old_item_file.exists():
            current_filename = old_item_file.stem
            if current_filename != new_filename:
                if new_item_file.exists() and new_item_file != old_item_file:
                    unique_filename = get_unique_filename(new_item_file, item_id)
                    new_item_file = column_dir / f"{unique_filename}.md"
                    new_filename = unique_filename

                if not safe_rename_file(old_item_file, new_item_file):
                    new_item_file = old_item_file

        # Extract all metadata from item_data, excluding content and title which are handled separately
        item_metadata = {
            key: value for key, value in item_data.items()
            if key not in ["title", "description"]  # title is passed separately, description becomes content
        }

        # Ensure required fields have defaults
        item_metadata.setdefault("created_at", now())
        item_metadata.setdefault("updated_at", now())

        try:
            save_item_with_metadata(new_item_file, title, content, item_metadata)
            logger.debug(f"Successfully saved item file: {new_item_file}")
        except Exception as e:
            logger.error(f"Failed to save item file {new_item_file}: {e}", exc_info=True)
            raise

    def delete_item_from_column(
        self, board_name: str, column_name: str, item_id: ItemId
    ) -> bool:
        board_dir = get_board_directory_path(self.boards_dir, board_name)
        column_dir = get_column_directory_path(board_dir, column_name)

        item_file = find_item_file_by_id(column_dir, item_id)
        if item_file and item_file.exists():
            return safe_delete_file(item_file)
        return False

    def move_item_between_columns(
        self,
        board_name: str,
        old_column_name: str,
        new_column_name: str,
        item_data: dict,
    ) -> bool:
        board_dir = get_board_directory_path(self.boards_dir, board_name)
        old_column_dir = get_column_directory_path(board_dir, old_column_name)
        new_column_dir = get_column_directory_path(board_dir, new_column_name)

        ensure_directory_exists(new_column_dir)

        item_id = item_data["id"]
        old_item_file = find_item_file_by_id(old_column_dir, item_id)

        if old_item_file and old_item_file.exists():
            item_data["updated_at"] = now()
            self.save_item_to_column(board_name, new_column_name, item_data)
            safe_delete_file(old_item_file)
            return True

        return False

    def save_column_metadata(
        self, board_name: str, column_name: str, column_data: dict
    ) -> None:
        board_dir = get_board_directory_path(self.boards_dir, board_name)
        column_dir = get_column_directory_path(board_dir, column_name)
        ensure_directory_exists(column_dir)
        column_metadata_file = column_dir / COLUMN_METADATA_FILENAME

        metadata = {
            "position": column_data.get("position"),
            "created_at": column_data.get("created_at", now()),
            "updated_at": column_data.get("updated_at", now()),
        }

        if column_data.get("limit") is not None:
            metadata["limit"] = column_data["limit"]

        save_column_metadata(column_metadata_file, column_name, metadata)

    def save_column_metadata_if_needed(
        self, board_name: str, column_data: dict
    ) -> None:
        needs_metadata = (
            column_data.get("position", 0) != 0 or column_data.get("limit") is not None
        )

        if needs_metadata:
            self.save_column_metadata(board_name, column_data["name"], column_data)

    def cleanup_column(
        self, board_name: str, column_name: str, current_item_ids: set[ItemId]
    ) -> None:
        board_dir = get_board_directory_path(self.boards_dir, board_name)
        column_dir = get_column_directory_path(board_dir, column_name)
        cleanup_column_files(column_dir, current_item_ids)

    def get_board_file_path(self, board_name: str) -> Path:
        board_dir = get_board_directory_path(self.boards_dir, board_name)
        return board_dir / BOARD_FILENAME

    def list_board_directories(self) -> list[Path]:
        if not self.boards_dir.exists():
            return []
        return [d for d in self.boards_dir.iterdir() if d.is_dir()]

File: mobile/src/infrastructure/storage/BoardPersistence.ts
/**
 * Board Persistence Layer
 * Low-level file operations for saving/loading board data
 * Ported from Python: src/infrastructure/storage/board_persistence.py
 */

import { FileSystemManager } from "./FileSystemManager";
import { MarkdownParser } from "./MarkdownParser";
import {
  findItemFileById,
  getBoardDirectoryPath,
  getColumnDirectoryPath,
  cleanupColumnFiles,
  getUniqueFilename,
} from "./FileOperations";
import { ItemId } from "../../core/types";
import { BOARD_FILENAME, COLUMN_METADATA_FILENAME } from "../../core/constants";
import { getTitleFilename } from "../../utils/stringUtils";
import { now } from "../../utils/dateUtils";

export interface ItemData {
  id: ItemId;
  title: string;
  description?: string;
  parent_id?: string | null;
  created_at?: Date;
  moved_in_progress_at?: Date | null;
  moved_in_done_at?: Date | null;
  worked_on_for?: string | null; // Format: "HH:MM"
  [key: string]: any; // Allow additional metadata
}

export interface ColumnData {
  id: string;
  name: string;
  position: number;
  limit?: number | null;
  created_at?: Date;
}

export class BoardPersistence {
  private fileSystem: FileSystemManager;
  private parser: MarkdownParser;
  private boardsDir: string;

  constructor(fileSystem: FileSystemManager, parser: MarkdownParser) {
    this.fileSystem = fileSystem;
    this.parser = parser;
    this.boardsDir = fileSystem.getBoardsDirectory();
  }

  /**
   * Save an item to a column directory
   */
  async saveItemToColumn(
    boardName: string,
    columnName: string,
    itemData: ItemData
  ): Promise<void> {
    try {
      const boardDir = getBoardDirectoryPath(this.boardsDir, boardName);
      const columnDir = getColumnDirectoryPath(boardDir, columnName);

      // Ensure column directory exists
      await this.fileSystem.ensureDirectoryExists(columnDir);

      const itemId = itemData.id;
      const title = itemData.title;
      const content = itemData.description || "";

      // Generate new filename: {id}-{title}.md (lowercase)
      const titlePart = getTitleFilename(title);
      const newFilename = `${itemId.toLowerCase()}-${titlePart}`;
      const newItemFile = `${columnDir}${newFilename}.md`;

      // Check if item already exists with different filename
      const oldItemFile = await findItemFileById(
        this.fileSystem,
        this.parser,
        columnDir,
        itemId
      );

      if (oldItemFile) {
        const currentFilename = this.getFileStem(oldItemFile);
        if (currentFilename !== newFilename) {
          // Need to rename - check for collisions
          const newFileExists = await this.fileSystem.fileExists(newItemFile);
          if (newFileExists && newItemFile !== oldItemFile) {
            // Collision - get unique filename
            const uniqueFilename = await getUniqueFilename(
              this.fileSystem,
              this.parser,
              newItemFile,
              itemId
            );
            const uniquePath = `${columnDir}${uniqueFilename}.md`;
            await this.fileSystem.renameFile(oldItemFile, uniquePath);
          } else {
            // Safe to rename
            await this.fileSystem.renameFile(oldItemFile, newItemFile);
          }
        }
      }

      // Extract metadata (exclude title and description which become content)
      const metadata: Record<string, any> = {};
      for (const [key, value] of Object.entries(itemData)) {
        if (key !== "title" && key !== "description") {
          metadata[key] = value;
        }
      }

      // Ensure created_at has a default
      if (!metadata.created_at) {
        metadata.created_at = now();
      }

      // Save the item with metadata
      const finalItemFile = oldItemFile || newItemFile;
      await this.parser.saveItemWithMetadata(finalItemFile, title, content, metadata);
    } catch (error) {
      throw new Error(
        `Failed to save item "${itemData.title}" to column "${columnName}": ${error}`
      );
    }
  }

  /**
   * Delete an item from a column
   */
  async deleteItemFromColumn(
    boardName: string,
    columnName: string,
    itemId: ItemId
  ): Promise<boolean> {
    try {
      const boardDir = getBoardDirectoryPath(this.boardsDir, boardName);
      const columnDir = getColumnDirectoryPath(boardDir, columnName);

      const itemFile = await findItemFileById(
        this.fileSystem,
        this.parser,
        columnDir,
        itemId
      );

      if (itemFile) {
        return await this.fileSystem.deleteFile(itemFile);
      }

      return false;
    } catch (error) {
      console.error(`Failed to delete item ${itemId} from column ${columnName}:`, error);
      return false;
    }
  }

  /**
   * Move an item between columns
   */
  async moveItemBetweenColumns(
    boardName: string,
    oldColumnName: string,
    newColumnName: string,
    itemData: ItemData
  ): Promise<boolean> {
    try {
      const boardDir = getBoardDirectoryPath(this.boardsDir, boardName);
      const oldColumnDir = getColumnDirectoryPath(boardDir, oldColumnName);
      const newColumnDir = getColumnDirectoryPath(boardDir, newColumnName);

      // Ensure new column directory exists
      await this.fileSystem.ensureDirectoryExists(newColumnDir);

      const itemId = itemData.id;

      // Find old item file
      const oldItemFile = await findItemFileById(
        this.fileSystem,
        this.parser,
        oldColumnDir,
        itemId
      );

      if (oldItemFile) {
        // Save to new column
        await this.saveItemToColumn(boardName, newColumnName, itemData);

        // Delete from old column
        await this.fileSystem.deleteFile(oldItemFile);
        return true;
      }

      return false;
    } catch (error) {
      console.error(
        `Failed to move item ${itemData.id} from ${oldColumnName} to ${newColumnName}:`,
        error
      );
      return false;
    }
  }

  /**
   * Save column metadata
   */
  async saveColumnMetadata(
    boardName: string,
    columnName: string,
    columnData: ColumnData
  ): Promise<void> {
    try {
      const boardDir = getBoardDirectoryPath(this.boardsDir, boardName);
      const columnDir = getColumnDirectoryPath(boardDir, columnName);

      await this.fileSystem.ensureDirectoryExists(columnDir);

      const columnMetadataFile = `${columnDir}${COLUMN_METADATA_FILENAME}`;

      const metadata: Record<string, any> = {
        position: columnData.position,
        created_at: columnData.created_at || now(),
      };

      if (columnData.limit !== undefined && columnData.limit !== null) {
        metadata.limit = columnData.limit;
      }

      await this.parser.saveColumnMetadata(columnMetadataFile, columnName, metadata);
    } catch (error) {
      throw new Error(`Failed to save column metadata for "${columnName}": ${error}`);
    }
  }

  /**
   * Clean up orphaned files in a column
   */
  async cleanupColumn(
    boardName: string,
    columnName: string,
    currentItemIds: Set<ItemId>
  ): Promise<void> {
    try {
      const boardDir = getBoardDirectoryPath(this.boardsDir, boardName);
      const columnDir = getColumnDirectoryPath(boardDir, columnName);

      await cleanupColumnFiles(this.fileSystem, this.parser, columnDir, currentItemIds);
    } catch (error) {
      console.error(`Failed to cleanup column ${columnName}:`, error);
    }
  }

  /**
   * Get the path to a board's kanban.md file
   */
  getBoardFilePath(boardName: string): string {
    const boardDir = getBoardDirectoryPath(this.boardsDir, boardName);
    return `${boardDir}${BOARD_FILENAME}`;
  }

  /**
   * List all board directories
   */
  async listBoardDirectories(): Promise<string[]> {
    try {
      const exists = await this.fileSystem.directoryExists(this.boardsDir);
      if (!exists) {
        return [];
      }

      return await this.fileSystem.listDirectories(this.boardsDir);
    } catch (error) {
      console.error(`Failed to list board directories:`, error);
      return [];
    }
  }

  /**
   * Get file stem (filename without extension)
   */
  private getFileStem(filePath: string): string {
    const parts = filePath.split("/");
    const filename = parts[parts.length - 1];
    return filename.replace(/\.md$/, "");
  }
}
